net/connUDP.go

```diff
@@ -208,88 +205,173 @@ func (c *UDPConn) Close() error {
+			return err
+		}
+	}
+	if err := p.SetMulticastHopLimit(multicastHopLimit); err != nil {
+		return err
+	}
+
+	var err error
+	if iface != nil || src != nil {
+		_, err = p.WriteTo(buffer, &ControlMessage{
+			Src:     pktSrc,
+			IfIndex: iface.Index,
+		}, raddr)
 	} else {
-		p = newPacketConnIPv6(ipv6.NewPacketConn(c.connection))
+		_, err = p.WriteTo(buffer, nil, raddr)
 	}
+	return err
+}
 
-	if err := p.SetMulticastInterface(&iface); err != nil {
-		return err
+func filterAddressesByNetwork(network string, ifaceAddrs []net.Addr) []net.Addr {
+	filtered := make([]net.Addr, 0, len(ifaceAddrs))
+	for _, srcAddr := range ifaceAddrs {
+		addrMask := srcAddr.String()
+		addr := strings.Split(addrMask, "/")[0]
+		if strings.Contains(addr, ":") && network == "udp4" {
+			continue
+		}
+		if !strings.Contains(addr, ":") && network == "udp6" {
+			continue
+		}
+		filtered = append(filtered, srcAddr)
 	}
-	if err := p.SetMulticastHopLimit(multicastHopLimit); err != nil {
+	return filtered
+
+}
+
+func convAddrsToIps(ifaceAddrs []net.Addr) []net.IP {
+	ips := make([]net.IP, 0, len(ifaceAddrs))
+	for _, addr := range ifaceAddrs {
+		addrMask := addr.String()
+		addr := strings.Split(addrMask, "/")[0]
+		ip := net.ParseIP(addr)
+		if ip != nil {
+			ips = append(ips, ip)
+		}
+	}
+	return ips
+}
+
+func (c *UDPConn) WriteMulticast(ctx context.Context, raddr *net.UDPAddr, buffer []byte, opts ...MulticastOption) error {
+	opt := MulticastOptions{
+		HopLimit: 1,
+	}
+	for _, o := range opts {
+		o.applyMC(&opt)
+	}
+	return c.writeMulticast(ctx, raddr, buffer, opt)
+}
+
+func (c *UDPConn) writeMulticastWithInterface(ctx context.Context, raddr *net.UDPAddr, buffer []byte, opt MulticastOptions) error {
+	if opt.Iface == nil && opt.IFaceMode == MulticastSpecificInterface {
+		return fmt.Errorf("invalid interface")
+	}
+	if opt.Source != nil {
+		return c.writeToAddr(opt.Iface, opt.Source, opt.HopLimit, raddr, buffer)
+	}
+	ifaceAddrs, err := opt.Iface.Addrs()
+	if err != nil {
 		return err
 	}
-	ip := net.ParseIP(addr)
-	if ip == nil {
-		return fmt.Errorf("cannot parse ip (%v) for iface %v", ip, iface.Name)
+	netType := "udp4"
+	if IsIPv6(raddr.IP) {
+		netType = "udp6"
 	}
-	if c.closed.Load() {
-		return ErrConnectionIsClosed
+	var errors []error
+	for _, ip := range convAddrsToIps(filterAddressesByNetwork(netType, ifaceAddrs)) {
+		opt.Source = &ip
+		err = c.writeToAddr(opt.Iface, opt.Source, opt.HopLimit, raddr, buffer)
+		if err != nil {
+			errors = append(errors, err)
+		}
 	}
-	_, err := p.WriteTo(buffer, &ControlMessage{
-		Src:     ip,
-		IfIndex: iface.Index,
-	}, raddr)
-	return err
+	if errors == nil {
+		return nil
+	}
+	if len(errors) == 1 {
+		return errors[0]
+	}
+	return fmt.Errorf("%v", errors)
 }
 
-func (c *UDPConn) WriteMulticast(ctx context.Context, raddr *net.UDPAddr, hopLimit int, buffer []byte) error {
+func (c *UDPConn) writeMulticast(ctx context.Context, raddr *net.UDPAddr, buffer []byte, opt MulticastOptions) error {
```

Comment:
SonarCloud complains that this method is too complex: 

> Refactor this method to reduce its Cognitive Complexity from 27 to the 15 allowed.Why is this an issue?

Reply:
Fixed